File: features/list/src/main/java/com/vp/list/GridPagingScrollListener.java
package com.vp.list;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class GridPagingScrollListener extends RecyclerView.OnScrollListener {
    private static final int PAGE_SIZE = 10;
    private final GridLayoutManager layoutManager;
    /* Should not be static as it's will keep a strong reference to the object(listener), and it's better like this(following The Android Profiler) - CE */
    private LoadMoreItemsListener loadMoreItemsListener;
    private boolean isLastPage = false;
    private boolean isLoading = false;

    GridPagingScrollListener(@NonNull final GridLayoutManager layoutManager) {
        this.layoutManager = layoutManager;
    }

    @Override
    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);
        if (shouldLoadNextPage()) {
            if(null != loadMoreItemsListener) {
                loadMoreItemsListener.loadMoreItems(getNextPageNumber());
            }
            else {
                // Empty listener
            }
        }
    }

    private boolean shouldLoadNextPage() {
        return isNotLoadingInProgress() && userScrollsToNextPage() && isNotFirstPage() && hasNextPage();
    }

    private boolean userScrollsToNextPage() {
        return (layoutManager.getChildCount() + layoutManager.findFirstVisibleItemPosition()) >= layoutManager.getItemCount();
    }

    private boolean isNotFirstPage() {
        return layoutManager.findFirstVisibleItemPosition() >= 0 && layoutManager.getItemCount() >= PAGE_SIZE;
    }

    private boolean hasNextPage() {
        return !isLastPage;
    }

    private boolean isNotLoadingInProgress() {
        return !isLoading;
    }

    private int getNextPageNumber() {
        return layoutManager.getItemCount() / PAGE_SIZE + 1;
    }

    public void setLoadMoreItemsListener(@Nullable final LoadMoreItemsListener loadMoreItemsListener) {
            this.loadMoreItemsListener = loadMoreItemsListener;
    }

    public void markLoading(boolean isLoading) {
        this.isLoading = isLoading;
    }

    public void markLastPage(boolean isLastPage) {
        this.isLastPage = isLastPage;
    }

    public interface LoadMoreItemsListener {
        void loadMoreItems(int page);
    }
}

File: features/list/src/main/java/com/vp/list/GridPagingScrollListener.kt
package com.vp.list

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.OnScrollListener

class GridPagingScrollListener internal constructor(
  private val layoutManager: GridLayoutManager
) : OnScrollListener() {
  private lateinit var loadMoreItemsListener: LoadMoreItemsListener
  private var isLastPage = false
  private var isLoading = false

  override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
    super.onScrolled(recyclerView, dx, dy)
    if (shouldLoadNextPage()) {
      loadMoreItemsListener.loadMoreItems(nextPageNumber)
    }
  }

  private fun shouldLoadNextPage() = isNotLoadingInProgress && userScrollsToNextPage() && isNotFirstPage && hasNextPage()

  private fun userScrollsToNextPage() = layoutManager.childCount + layoutManager.findFirstVisibleItemPosition() >= layoutManager.itemCount

  private val isNotFirstPage: Boolean
    get() = layoutManager.findFirstVisibleItemPosition() >= 0 && layoutManager.itemCount >= PAGE_SIZE

  private fun hasNextPage() = !isLastPage

  private val isNotLoadingInProgress: Boolean
    get() = !isLoading

  private val nextPageNumber: Int
    get() = layoutManager.itemCount / PAGE_SIZE + 1

  fun setLoadMoreItemsListener(loadMoreItemsListener: LoadMoreItemsListener) {
    this.loadMoreItemsListener = loadMoreItemsListener
  }

  fun markLoading(isLoading: Boolean) {
    this.isLoading = isLoading
  }

  fun markLastPage(isLastPage: Boolean) {
    this.isLastPage = isLastPage
  }

  interface LoadMoreItemsListener {
    fun loadMoreItems(page: Int)
  }

  companion object {
    private const val PAGE_SIZE = 10
  }
}